### PR TITLE
feat(compiler): add "original" placeholder value on extracted XMB

### PR DESCRIPTION
**Pull request by Goodwine**
Update XMB placeholders(`<ph>`) to include the original value on top of an
example. Placeholders can by definition have one example(`<ex>`) tag and a
text node. The text node is used by TC as the "original" value from the
placeholder, while the example should represent a dummy value.
For example: `<ph name="PET"><ex>Gopher</ex>{{ petName }}</ph>`.
This change makes sure that we have the original text, but it **DOES NOT**
make sure that the example is correct. The example has the same wrong
behavior of showing the interpolation text rather than a useful example.

No breaking changes, but tools that depend on the previous behavior and
don't consider the full XMB definition may fail to parse the XMB.
Fixes b/72565847

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[x] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[x] Other... Please describe: XMB extraction
```

## What is the current behavior?

A message like `<foo>Name: {{yourName}}</foo>` would generate this xmb message:
```xml
<msg id=123>Name: <ph name="INTERPOLATION"><ex>{{yourName}}</ex></ph></msg>
```

Issue Number: http://b/72565847

## What is the new behavior?
A message like `<foo>Name: {{yourName}}</foo>` would generate this xmb message:
```xml
<msg id=123>Name: <ph name="INTERPOLATION"><ex>{{yourName}}</ex>{{yourName}}</ph></msg>
```
The `<ph>` PCDATA (text node) is used by TC for some validations explained further in the internal bug.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

Tools that depend on the current (wrong) implementation may fail if they don't take the `#PCDATA` definition into account.

## Other information
Ideally the example should be something like `<ex>Carlos</ex>`, but fixing the example is out of scope because there is no way to provide one.

### CODE REVIEW

**Comment by Goodwine** [comment]
Note: These bazel tests were already failing

//packages/compiler-cli/test/compliance:compliance
//packages/core/test/bundling/hello_world:symbol_test
//packages/core/test/bundling/hello_world:test
//packages/core/test/bundling/todo:symbol_test
//packages/core/test/bundling/todo:test

File: packages/compiler-cli/test/BUILD.bazel [hunk 1]
```diff
@@ -43,6 +43,7 @@ jasmine_node_test(
     ],
     deps = [
         ":extract_i18n_lib",
+        "//packages/common:npm_package",
```
Comment (vicb) [inline packages/compiler-cli/test/BUILD.bazel]: Could you explain why this change is required ? Thx

**Comment by vicb** [comment]
@Goodwine Thanks for the PR. I'll do my best to review it by Thursday.

File: packages/compiler-cli/test/BUILD.bazel (hunk 1 above)
Comment (Goodwine) [reply packages/compiler-cli/test/BUILD.bazel]: CommonModule is needed to process ICU syntax, otherwise angular complains that it doesn't know how to handle ICU. The module is imported in a generated "module" test file similar to `@NgModule`, but `CommonModule` is in a different package.

File: packages/compiler/src/i18n/serializers/xmb.ts [hunk 2]
```diff
@@ -115,30 +115,36 @@ class _Visitor implements i18n.Visitor {
   }
 
   visitTagPlaceholder(ph: i18n.TagPlaceholder, context?: any): xml.Node[] {
-    const startEx = new xml.Tag(_EXEMPLE_TAG, {}, [new xml.Text(`<${ph.tag}>`)]);
-    const startTagPh = new xml.Tag(_PLACEHOLDER_TAG, {name: ph.startName}, [startEx]);
+    const originalStartTag = new xml.Text(`<${ph.tag}>`);
```
Comment (vicb) [inline packages/compiler/src/i18n/serializers/xmb.ts]: nit: rename to `startTagAsText` or something along those line

File: packages/compiler/src/i18n/serializers/xmb.ts [hunk 3]
```diff
@@ -115,30 +115,36 @@ class _Visitor implements i18n.Visitor {
...
+    const startTagPh =
+        new xml.Tag(_PLACEHOLDER_TAG, {name: ph.startName}, [startEx, originalStartTag]);
     if (ph.isVoid) {
       // void tags have no children nor closing tags
       return [startTagPh];
     }
 
-    const closeEx = new xml.Tag(_EXEMPLE_TAG, {}, [new xml.Text(`</${ph.tag}>`)]);
-    const closeTagPh = new xml.Tag(_PLACEHOLDER_TAG, {name: ph.closeName}, [closeEx]);
+    const originalCloseTag = new xml.Text(`</${ph.tag}>`);
```
Comment (vicb) [inline packages/compiler/src/i18n/serializers/xmb.ts]: nit: `closeTagAsText`

File: packages/compiler/src/i18n/serializers/xmb.ts [hunk 4]
```diff
@@ -115,30 +115,36 @@ class _Visitor implements i18n.Visitor {
...
   }
 
   visitIcuPlaceholder(ph: i18n.IcuPlaceholder, context?: any): xml.Node[] {
-    const exTag = new xml.Tag(_EXEMPLE_TAG, {}, [
-      new xml.Text(
-          `{${ph.value.expression}, ${ph.value.type}, ${Object.keys(ph.value.cases).map((value: string) => value + ' {...}').join(' ')}}`)
-    ]);
-    return [new xml.Tag(_PLACEHOLDER_TAG, {name: ph.name}, [exTag])];
+    const originalIcu = new xml.Text(
+        `{${ph.value.expression}, ${ph.value.type}, ${Object.keys(ph.value.cases).map((value: string) => value + ' {...}').join(' ')}}`);
```
Comment (vicb) [inline packages/compiler/src/i18n/serializers/xmb.ts]: nit: would you mind extracting the `${}` to local `const`s ?
(I know this has not been introduced by you but it would help readability anyway)

File: packages/compiler/test/i18n/integration_xmb_xtb_spec.ts [hunk 5]
```diff
@@ -91,32 +91,32 @@ const XTB = `
...
   <msg id="5525133077318024839"><source>file.ts:14</source>on not translatable node</msg>
   <msg id="2174788525135228764"><source>file.ts:14</source>&lt;b&gt;bold&lt;/b&gt;</msg>
   <msg id="8670732454866344690"><source>file.ts:15</source>on translatable node</msg>
-  <msg id="4593805537723189714"><source>file.ts:20</source><source>file.ts:37</source>{VAR_PLURAL, plural, =0 {zero} =1 {one} =2 {two} other {<ph name="START_BOLD_TEXT"><ex>&lt;b&gt;</ex></ph>many<ph name="CLOSE_BOLD_TEXT"><ex>&lt;/b&gt;</ex></ph>} }</msg>
+  <msg id="4593805537723189714"><source>file.ts:20</source><source>file.ts:37</source>{VAR_PLURAL, plural, =0 {zero} =1 {one} =2 {two} other {<ph name="START_BOLD_TEXT"><ex>&lt;b&gt;</ex>&lt;b&gt;</ph>many<ph name="CLOSE_BOLD_TEXT"><ex>&lt;/b&gt;</ex>&lt;/b&gt;</ph>} }</msg>
   <msg id="703464324060964421"><source>file.ts:22,24</source>
-        <ph name="ICU"><ex>{sex, select, male {...} female {...} other {...}}</ex></ph>
+        <ph name="ICU"><ex>male, female, other</ex>{sex, select, male {...} female {...} other {...}}</ph>
```
Comment (vicb) [inline packages/compiler/test/i18n/integration_xmb_xtb_spec.ts]: is changing the `<ex>` a requirement from the TC

File: packages/compiler/test/i18n/integration_xmb_xtb_spec.ts [hunk 6]
```diff
@@ -91,32 +91,32 @@ const XTB = `
...
   <msg id="5539162898278769904" meaning="different meaning"><source>file.ts:7</source>nested</msg>
-  <msg id="3780349238193953556"><source>file.ts:9</source><source>file.ts:10</source><ph name="START_ITALIC_TEXT"><ex>&lt;i&gt;</ex></ph>with placeholders<ph name="CLOSE_ITALIC_TEXT"><ex>&lt;/i&gt;</ex></ph></msg>
-  <msg id="5415448997399451992"><source>file.ts:11</source><ph name="START_TAG_DIV"><ex>&lt;div&gt;</ex></ph>with <ph name="START_TAG_DIV"><ex>&lt;div&gt;</ex></ph>nested<ph name="CLOSE_TAG_DIV"><ex>&lt;/div&gt;</ex></ph> placeholders<ph name="CLOSE_TAG_DIV"><ex>&lt;/div&gt;</ex></ph></msg>
+  <msg id="3780349238193953556"><source>file.ts:9</source><source>file.ts:10</source><ph name="START_ITALIC_TEXT"><ex>&lt;i&gt;</ex>&lt;i&gt;</ph>with placeholders<ph name="CLOSE_ITALIC_TEXT"><ex>&lt;/i&gt;</ex>&lt;/i&gt;</ph></msg>
+  <msg id="5415448997399451992"><source>file.ts:11</source><ph name="START_TAG_DIV"><ex>&lt;div&gt;</ex>&lt;div&gt;</ph>with <ph name="START_TAG_DIV"><ex>&lt;div&gt;</ex>&lt;div&gt;</ph>nested<ph name="CLOSE_TAG_DIV"><ex>&lt;/div&gt;</ex>&lt;/div&gt;</ph> placeholders<ph name="CLOSE_TAG_DIV"><ex>&lt;/div&gt;</ex>&lt;/div&gt;</ph></msg>
```
Comment (vicb) [inline packages/compiler/test/i18n/integration_xmb_xtb_spec.ts]: `<div>` looks wrong here - is it github transforming encoded entities ?

**Comment by jermowery** [comment]
+100,000,000 this will eliminate an enormous annoyance

File: packages/compiler-cli/test/BUILD.bazel (hunk 1 above)
Comment (vicb) [reply packages/compiler-cli/test/BUILD.bazel]: Why not importing `//packages/common` instead ?

File: packages/compiler/src/i18n/serializers/xmb.ts (hunk 2 above)
Comment (vicb) [reply packages/compiler/src/i18n/serializers/xmb.ts]: Please also add a comment explaining that the TC uses this extra text node (for all placeholder types).

**Review by vicb — APPROVED** [review]
Thank you very much for this PR.
Could you please address the minor comments ?

File: packages/compiler-cli/test/BUILD.bazel (hunk 1 above)
Comment (Goodwine) [reply packages/compiler-cli/test/BUILD.bazel]: I copied it from another rule, when changing it to `//packages/common`, bazel fails the test `//packages/compiler-cli/test:extract_i18n` because "error TS2307: Cannot find module '@angular/common'."

File: packages/compiler/src/i18n/serializers/xmb.ts (hunk 4 above)
Comment (Goodwine) [reply packages/compiler/src/i18n/serializers/xmb.ts]: will do

File: packages/compiler/test/i18n/integration_xmb_xtb_spec.ts (hunk 5 above)
Comment (Goodwine) [reply packages/compiler/test/i18n/integration_xmb_xtb_spec.ts]: it's not, `<ex>` is used by translators to know what kind of values to expect, I figured it would make sense to only list the options instead of the icu's text representation

File: packages/compiler/test/i18n/integration_xmb_xtb_spec.ts (hunk 6 above)
Comment (Goodwine) [reply packages/compiler/test/i18n/integration_xmb_xtb_spec.ts]: This is the XML library when doing ``new xml.Text(`<${ph.tag}>`);``, it's WAI

**Comment by Goodwine** [comment]
Thanks for the review, I've addressed the comments, PTAL

File: packages/compiler/test/i18n/integration_xmb_xtb_spec.ts (hunk 5 above)
Comment (vicb) [reply packages/compiler/test/i18n/integration_xmb_xtb_spec.ts]: I would prefer not to change this one in this PR, Coud you please revert this change.

**Comment by vicb** [comment]
Could you please revert the `<ex>` change - everything else is great.
Thanks !

**Comment by vicb** [comment]
http://test/OCL:206611829:BASE:206611877:1532973875793:3559563b

**Comment by vicb** [comment]
merge-assistance: missing pullapprove but the change is only about a symbol that has been renamed (to fix a typo). Please merge. Thanks.

**Review by IgorMinar — APPROVED** [review]
